homeassistant/components/tplink/entity.py

```diff
@@ -57,24 +95,189 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
     return _async_wrap
 
 
-class CoordinatedTPLinkEntity(CoordinatorEntity[TPLinkDataUpdateCoordinator]):
+class CoordinatedTPLinkEntity(CoordinatorEntity[TPLinkDataUpdateCoordinator], ABC):
     """Common base class for all coordinated tplink entities."""
 
     _attr_has_entity_name = True
 
     def __init__(
-        self, device: Device, coordinator: TPLinkDataUpdateCoordinator
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        feature: Feature | None = None,
+        parent: Device | None = None,
     ) -> None:
-        """Initialize the switch."""
+        """Initialize the entity."""
         super().__init__(coordinator)
         self.device: Device = device
-        self._attr_unique_id = device.device_id
+        self._feature = feature
         self._attr_device_info = DeviceInfo(
-            connections={(dr.CONNECTION_NETWORK_MAC, device.mac)},
             identifiers={(DOMAIN, str(device.device_id))},
             manufacturer="TP-Link",
             model=device.model,
             name=device.alias,
             sw_version=device.hw_info["sw_ver"],
             hw_version=device.hw_info["hw_ver"],
         )
+
+        if parent is not None:
+            self._attr_device_info["via_device"] = (DOMAIN, parent.device_id)
+        else:
+            self._attr_device_info["connections"] = {
+                (dr.CONNECTION_NETWORK_MAC, device.mac)
+            }
+
+        # The rest of the initialization takes care of setting a proper unique_id
+        # This is transitional and will become cleaner as future platforms get converted.
+
+        # If the unique id is already defined, we don't need to do anything.
+        if self._attr_unique_id is not None:
+            return
+
```

Comment:
Its a bit unexpected that the check is based on unique id and not something more deterministic

Reply:
Yeah, I'll find a better solution for that, likely just by adding a boolean flag that is set to true by the specialized platforms.

Reply:
Moved all unique_id logic into `_get_unique_id` of the base entity class to have it in a single place.